Match making API

bjarkehs opened the pull request:
Closes #14 

This PR adds a match making API. The idea is that players can queue up and then when there's enough players it'll create a pending match. Players then have 30 seconds to accept the invitation before they are removed from queue - if they don't respond. If a player declines then they are also removed from queue. The players, who accepted, are placed back in the queue.

If all players accept, then a temporary match is created.
Future enhancement would be to allow players to vote for a how teams are assigned: Random or ranked.

A temporary match can be cancelled or the score can be entered for the temporary match and it becomes an actual match.

This temporary match model and API could also be used for other things, such as scanners.

We should show the temporary matches as "Matches in progress" on the website, so people can see who are playing. Further enhancement would be to configure available tables and set up a queueing system for when there are more matches than tables (in the case of foosball).

To be done in this PR:

- [x] Add temporary match (or similar name)
- [x] Pending match -> temporary match
- [ ] ~~Voting on team setup~~
- [x] Migration(s)
- [x] API for temporary matches
- [x] Frontend code
- [ ] ~~Push-model instead of pull (SignalR?)~~

UPDATE:
Skipping the two unfinished for now. They should/could be implemented later.

bjarkehs committed:
WIP

bjarkehs committed:
Implement APIs for accepting pending matches

bjarkehs committed:
Add background service for match making

lajtmaN commented on backend-dotnet/MMRProject.Api/DTOs/PendingMatchDto.cs:
```diff
@@ -0,0 +1,9 @@
+
+public record PendingMatchDto
+{
+    public long Id { get; set; }
```
Er lidt øv over at vi bruger long til id'er. 

Det næstbedste er at vi ikke udstiller det som tal i vores api, men blot som en string, så det bliver nemmere for os at gå over til uuid/ulid/... senere

bjarkehs replied:
Enig. Jeg har rettet det nu. Det var kun midlertidigt jeg havde det som `long`, mens jeg lige lavede logik 😄

lajtmaN commented on backend-dotnet/MMRProject.Api/Services/MatchMakingService.cs:
```diff
@@ -0,0 +1,234 @@
+        return new MatchMakingQueueStatus
+        {
+            PlayersInQueue = queuedPlayers.Count,
+            UserInQueue = queuedPlayers.Any(x => x.User.IdentityUserId == identityUserId)
```
Rename til `IsUserInQueue` - var først da jeg nåede til implementering at jeg forstod det var det som vi gemte

bjarkehs replied:
Done

bjarkehs committed:
Add API for handling active matches

lajtmaN reviewed (commented):
Bliver mega fedt at få dette lavet 👏🏼

bjarkehs committed:
Rename field for whether user is queued up

bjarkehs committed:
Change long ids of new entities to UUIDv7

bjarkehs committed:
Transition from pending to active match

bjarkehs committed:
Add migration

bjarkehs committed:
Fix issue with background service unable to start

bjarkehs committed:
Remove warning about global filters for user

bjarkehs committed:
Fix frontend openapi generation

bjarkehs committed:
Add endpoints to bruno

bjarkehs committed:
Fix model names in missing places

lajtmaN committed:
Added matchmaking route in frontend

lajtmaN committed:
Merge branch 'bjarkehs/match-making' of github.com:office-rivals/mmr-project into bjarkehs/match-making

bjarkehs committed:
Fix more model type issues

lajtmaN committed:
Make sure user has claimed a profile

lajtmaN committed:
Merge branch 'bjarkehs/match-making' of github.com:office-rivals/mmr-project into bjarkehs/match-making

lajtmaN committed:
Started on matchmaking card that displays current status

lajtmaN committed:
Updated queue status message

bjarkehs committed:
Fix more broken model references

lajtmaN committed:
Added pending match indicator

bjarkehs committed:
Initial work on active matches in UI

bjarkehs committed:
Add method for cancelling active match

bjarkehs committed:
Add pending match to queue indicator

bjarkehs committed:
Tweak logic for queue status

bjarkehs committed:
Fix enum mapping

bjarkehs committed:
Attempt to handle match accepted state

bjarkehs committed:
Improve responsiveness of queue indicator

bjarkehs committed:
Improve responsiveness of matchmaking page

bjarkehs committed:
Reduce amount of backend calls when not in queue

bjarkehs committed:
Add notification sound

bjarkehs committed:
Show players in queue on indicator

bjarkehs committed:
Fix submitting active matches

bjarkehs committed:
Add an extra second to expires at

bjarkehs committed:
Merge remote-tracking branch 'origin/main' into bjarkehs/match-making

madsthom reviewed (approved):
LGTM GIANT PR

lajtmaN committed:
Fix animation on countdown

Tailwind cannot generate a custom class on build time when it contains variables, thus the variable is applied using style to override
